File: plugins/apps/qwenpaw-creator/ui/src/contracts/creator/sessions.ts
export type CreatorSessionStatus =
  | "IDLE"
  | "RUNNING"
  | "WAITING_RUNTIME"
  | "INTERRUPT_REQUESTED"
  | "WAITING_USER_INPUT"
  | "WAITING_EXECUTION_AUTH"
  | "PENDING_REVIEW"
  | "RESUMING"
  | "CANCELLED"
  | "ERROR";

export type CreatorProgressPhase =
  | "source_ingest"
  | "source_intelligence"
  | "creative_strategy"
  | "visual_development"
  | "timeline_edit"
  | "timeline_render"
  | "post_production"
  | "review"
  | "completed";

export interface CreatorTaskProgress {
  goalId?: string | null;
  phase: CreatorProgressPhase;
  label: string;
  latestMilestone?: string | null;
  completed?: number | null;
  total?: number | null;
  timelineId?: string | null;
  elementId?: string | null;
  sourceEventSeq: number;
  updatedAt: string;
}

export interface AgentStatusBadge {
  kind: string;
  label: string;
  count?: number | null;
  actionTarget?: string | null;
}

export interface AgentStatusBarView {
  progress: CreatorTaskProgress;
  activity?: {
    runningTaskCount?: number;
    [key: string]: unknown;
  } | null;
  badges: AgentStatusBadge[];
}

export type CreatorContentPart =
  | { type: "text"; text: string }
  | {
      type: "image_url";
      image_url: { url: string };
      attachment?: { assetVersionRef?: string; mediaType?: string };
    }
  | {
      type: "video_url";
      video_url: { url: string };
      fps?: number;
      max_frames?: number;
      attachment?: { assetVersionRef?: string; mediaType?: string };
    }
  | { type: "audio" | "document"; attachment: Record<string, unknown> };

export interface CreatorSessionView {
  id: string;
  projectId: string;
  status: CreatorSessionStatus;
  activeGoalId?: string;
  lastMessageSeq: number;
  lastConsumedMessageSeq: number;
  lastEventSeq: number;
  error?: {
    code: string;
    message: string;
    retryable: boolean;
    details?: Record<string, unknown>;
  } | null;
}

export interface CreatorMessage {
  messageId: string;
  messageSeq: number;
  role: "system" | "user" | "assistant" | "tool";
  content: CreatorContentPart[];
  source?: string;
  metadata: Record<string, unknown>;
  createdAt: string;
}

export interface CreatorConversation {
  conversationId: string;
  title: string;
  isDefault: boolean;
  createdAt: string;
}

export interface ConversationPage {
  items: CreatorConversation[];
  nextCursor?: number | null;
}

export interface MessagePage {
  items: CreatorMessage[];
  nextAfter?: number | null;
  /** Backward cursor: smallest seq of this page when older history exists. */
  nextBefore?: number | null;
}

export interface CreatorSessionResponse {
  session: CreatorSessionView;
  agentStatusBar: AgentStatusBarView;
}

export interface SendCreatorMessageRequest {
  clientMessageId: string;
  creatorSessionId?: string;
  conversationId: string;
  message?: string;
  content?: CreatorContentPart[];
  assetVersionRefs?: string[];
  context?: Record<string, unknown>;
}

export interface MessageAccepted {
  messageSeq: number;
  eventSeq: number;
  classification:
    | "read_only_question"
    | "mutation_instruction"
    | "review_comment"
    | "review_revise"
    | "workspace_command";
  appendState: "appended" | "queued_until_message_boundary";
  creatorSessionId: string;
  conversationId: string;
}

export interface ConversationCreated {
  conversationId: string;
  creatorSessionId: string;
  title: string;
  createdAt: string;
}

File: plugins/apps/qwenpaw-creator/backend/schemas/sessions.py
# -*- coding: utf-8 -*-
from __future__ import annotations

from typing import Any, Literal

from pydantic import Field

from .common import ContentPart, StrictModel
from .status import AgentStatusBarView


CreatorSessionStatus = Literal[
    "IDLE",
    "RUNNING",
    "WAITING_RUNTIME",
    "INTERRUPT_REQUESTED",
    "WAITING_USER_INPUT",
    "WAITING_EXECUTION_AUTH",
    "PENDING_REVIEW",
    "RESUMING",
    "CANCELLED",
    "ERROR",
]


class CreatorSessionView(StrictModel):
    id: str
    project_id: str = Field(alias="projectId")
    status: CreatorSessionStatus
    active_goal_id: str | None = Field(None, alias="activeGoalId")
    last_message_seq: int = Field(alias="lastMessageSeq", ge=0)
    last_consumed_message_seq: int = Field(
        alias="lastConsumedMessageSeq",
        ge=0,
    )
    last_event_seq: int = Field(alias="lastEventSeq", ge=0)
    error: dict[str, Any] | None = None


class CreatorSessionResponse(StrictModel):
    session: CreatorSessionView
    agent_status_bar: AgentStatusBarView = Field(alias="agentStatusBar")


class CreatorConversationView(StrictModel):
    conversation_id: str = Field(alias="conversationId")
    title: str
    is_default: bool = Field(alias="isDefault")
    created_at: str = Field(alias="createdAt")


class ConversationPage(StrictModel):
    items: list[CreatorConversationView]
    next_cursor: int | None = Field(None, alias="nextCursor")


class ConversationCreated(StrictModel):
    conversation_id: str = Field(alias="conversationId")
    creator_session_id: str = Field(alias="creatorSessionId")
    title: str
    created_at: str = Field(alias="createdAt")


class CreatorMessageView(StrictModel):
    message_id: str = Field(alias="messageId")
    message_seq: int = Field(alias="messageSeq", ge=1)
    role: Literal["system", "user", "assistant", "tool"]
    content: list[dict[str, Any]]
    source: str | None = None
    metadata: dict[str, Any]
    created_at: str = Field(alias="createdAt")


class MessagePage(StrictModel):
    items: list[CreatorMessageView]
    next_after: int | None = Field(None, alias="nextAfter")
    # Backward-pagination cursor: the smallest seq of this page when older
    # history exists (``before``/``tail`` requests only).
    next_before: int | None = Field(None, alias="nextBefore")


class CreatorMessageRequest(StrictModel):
    client_message_id: str = Field(alias="clientMessageId")
    creator_session_id: str | None = Field(None, alias="creatorSessionId")
    conversation_id: str = Field(alias="conversationId")
    message: str | None = None
    content: list[ContentPart] = Field(default_factory=list)
    asset_version_refs: list[str] = Field(
        default_factory=list,
        alias="assetVersionRefs",
    )
    context: dict[str, Any] = Field(default_factory=dict)


class CreatorMessageAccepted(StrictModel):
    message_seq: int = Field(alias="messageSeq")
    event_seq: int = Field(alias="eventSeq")
    classification: Literal[
        "read_only_question",
        "mutation_instruction",
        "review_comment",
        "review_revise",
        "workspace_command",
    ]
    append_state: Literal["appended", "queued_until_message_boundary"] = Field(
        alias="appendState",
    )
    creator_session_id: str = Field(alias="creatorSessionId")
    conversation_id: str = Field(alias="conversationId")
